android/src/main/kotlin/com/example/flutter_gamepad/FlutterGamepadPlugin.kt

```diff
@@ -1,25 +1,91 @@
 import io.flutter.plugin.common.MethodCall
 import io.flutter.plugin.common.MethodChannel
 import io.flutter.plugin.common.MethodChannel.MethodCallHandler
 import io.flutter.plugin.common.MethodChannel.Result
 import io.flutter.plugin.common.PluginRegistry.Registrar
+import io.flutter.plugin.editing.TextInputPlugin
+import java.lang.reflect.Field
 
-class FlutterGamepadPlugin: MethodCallHandler {
-  companion object {
-    @JvmStatic
-    fun registerWith(registrar: Registrar) {
-      val channel = MethodChannel(registrar.messenger(), "flutter_gamepad")
-      channel.setMethodCallHandler(FlutterGamepadPlugin())
+/**
+ * An extension of Flutter's AndroidTouchProcessor that delegates MotionEvents to the GamepadStreamHandler.
+ */
+class GamepadAndroidTouchProcessor(renderer: FlutterRenderer) : AndroidTouchProcessor(renderer) {
+    override fun onGenericMotionEvent(event: MotionEvent): Boolean {
+        return GamepadStreamHandler.processMotionEvent(event) || super.onGenericMotionEvent(event)
+    }
+}
+
+/**
+ * An extension of Flutter's AndroidKeyProcessor that delegates KeyEvents to the GamepadStreamHandler.
+ */
+class GamepadAndroidKeyProcessor(keyEventChannel: KeyEventChannel, textInputPlugin: TextInputPlugin) : AndroidKeyProcessor(keyEventChannel, textInputPlugin) {
+    override fun onKeyDown(keyEvent: KeyEvent) {
+        val handled = GamepadStreamHandler.processKeyDownEvent(keyEvent)
+        if (!handled) {
+            super.onKeyDown(keyEvent)
+        }
+    }
+
+    override fun onKeyUp(keyEvent: KeyEvent) {
+        val handled = GamepadStreamHandler.processKeyUpEvent(keyEvent)
+        if (!handled) {
+            super.onKeyDown(keyEvent)
+        }
+    }
+}
+
+/**
+ * The flutter_gamepad plugin class that is registered with the framework.
+ */
+class FlutterGamepadPlugin : MethodCallHandler {
+    companion object {
+        @JvmStatic
+        fun registerWith(registrar: Registrar) {
+            val methodChannel = MethodChannel(registrar.messenger(), "com.rainway.flutter_gamepad/methods")
+            methodChannel.setMethodCallHandler(FlutterGamepadPlugin())
+            val eventChannel = EventChannel(registrar.messenger(), "com.rainway.flutter_gamepad/events")
+            eventChannel.setStreamHandler(GamepadStreamHandler)
+
+            val view: FlutterView = registrar.view()
+            fun viewField(name: String): Field {
+                val field = FlutterView::class.java.getDeclaredField(name)
+                field.isAccessible = true
+                return field
+            }
+
+            // Hack: swap in a new AndroidTouchProcessor.
+            val touchProcessorField = viewField("androidTouchProcessor")
+            val rendererField = viewField("flutterRenderer")
+            val renderer = rendererField.get(view) as FlutterRenderer
+            val touchProcessor = GamepadAndroidTouchProcessor(renderer)
+            touchProcessorField.set(view, touchProcessor)
```

Comment:
The "Hack" comment seems to suggest that this solution is tenuous in that it could possibly break if Flutter changes something about how it is implemented. For example if the "androidTouchProcessor" field changes names perhaps? If something like that were to happen would this code throw an uncaught exception? If so, should we try to catch the exception here and deal with it gracefully?

Reply:
Yes — it's relying on the presence of a (private!) field with that name in a class in the Flutter engine, which could change in a future version.

This patching step is crucial to the working of the plugin: if it goes wrong, then the plug-in can't possibly do its job (because we will have no access to key/motion events). In a sense, crashing feels like the “right” thing to do, then: users of the plug-in will make an issue here on GitHub, and it will be up to us to fix and accommodate the new Flutter version.

Reply:
Isn't this what [RawKeyboard ](https://api.flutter.dev/flutter/services/RawKeyboard-class.html) is for in combination with [InputDevice](https://developer.android.com/reference/android/view/InputDevice) using a `SOURCE_GAMEPAD` 


https://api.flutter.dev/flutter/services/RawKeyEventDataAndroid-class.html

Reply:
Yes, you (as a Flutter app developer) can use a RawKeyboard Flutter widget to get gamepad button events.

However, at the time of writing, that works only on Android, and only buttons are supported: the motion events from thumbsticks [are discarded](https://github.com/flutter/engine/blob/3bcd6a8f7b360cbdfe527b5e21eaca108c24f31f/shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java#L151-L156) by AndroidTouchProcessor. (The events also don't seem to have a `deviceId` field, only `vendorId` and `productId`, so theoretically you couldn't tell two identical Dualshocks connected to the device apart. This is kind of a nitpick.)

If we want to offer a plug-in that has the same API on both iOS and Android, i.e. if we want to put button events onto the `flutter_gamepad` event stream, then we need the trick, because Flutter widgets are immaterial to us at the plug-in level. So RawKeyboard's existence does not help us much.

Reply:
Unrelated to this PR, but just had the thought that it might be a good idea to specify on this https://pub.dev/packages/flutter_gamepad that it is tested and works with specific versions of Flutter and that it is potentially not compatible with versions that haven't been tested.

Also I'm curious, if the flutter_gamepad plugin did cause a crash due to incompatibility with the version of Flutter that the Flutter app developer is using it with, would it be obvious to the developer why/where it crashed (assuming the developer has no background in platform specific development so doing something like running the app through XCode is too foreign for them to try)? In other words, if it were to crash after said app developer did a "flutter run" on their app, would they be able to tell that it is the gamepad_plugin that is malfunctioning?